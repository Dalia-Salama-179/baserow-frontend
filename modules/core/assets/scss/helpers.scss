.clearfix::after {
  clear: both;
  content: '';
  display: table;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.background-white {
  background-color: $white;
}

.color-primary {
  color: $color-primary-500 !important;
}

.color-success {
  color: $color-success-500 !important;
}

.color-warning {
  color: $color-warning-500 !important;
}

.color-error {
  color: $color-error-500 !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottom-4 {
  margin-bottom: 40px !important;
}

.margin-left-auto {
  margin-left: auto !important;
}

.margin-left-0 {
  margin-left: 0 !important;
}

.margin-left-1 {
  margin-left: 8px !important;
}

.margin-left-2 {
  margin-left: 16px !important;
}

.margin-right-auto {
  margin-right: auto !important;
}

.margin-right-0 {
  margin-right: 0 !important;
}

.margin-right-1 {
  margin-right: 8px !important;
}

.margin-right-2 {
  margin-right: 16px !important;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.forced-text-overflow-initial {
  text-overflow: initial !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}
