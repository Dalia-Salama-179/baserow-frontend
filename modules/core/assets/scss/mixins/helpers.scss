@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if ($top != null and $right == null and $bottom == null and $left == null) {
    top: $top;
    right: $top;
    bottom: $top;
    left: $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if ($top != null and $right != null and $bottom == null and
  $left == null) {
    top: $top;
    right: $right;
    bottom: $top;
    left: $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top != null {
      top: $top;
    }

    @if $right != null {
      right: $right;
    }

    @if $bottom != null {
      bottom: $bottom;
    }

    @if $left != null {
      left: $left;
    }
  }
}
