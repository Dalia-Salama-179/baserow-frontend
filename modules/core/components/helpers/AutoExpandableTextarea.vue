<template>
  <textarea
    ref="inputTextArea"
    :value="value"
    :placeholder="placeholder"
    class="input auto-expandable-textarea"
    :style="{
      height: textBoxSize + 'px',
      overflow: textBoxOverflow,
    }"
    rows="1"
    @input="$emit('input', $event.target.value)"
    @click="$emit('click', $event)"
    @keyup="$emit('keyup', $event)"
    @keydown="$emit('keydown', $event)"
    @blur="$emit('blur', $event)"
  />
</template>
<script>
/**
 * A blank textarea with no surrounding box styling (see AutoExpandableTextareaInput if
 * you want a nice styled box around this) which will automatically expand given the
 * users input to the maximum number of rows specified by the maxRows prop.
 */
export default {
  name: 'AutoExpandableTextarea',
  props: {
    value: {
      required: true,
      type: String,
    },
    loading: {
      required: false,
      default: false,
      type: Boolean,
    },
    placeholder: {
      required: false,
      type: String,
      default: '',
    },
    maxRows: {
      required: false,
      type: Number,
      default: 4,
    },
    startingRows: {
      required: false,
      type: Number,
      default: 1,
    },
  },
  data() {
    return {
      numTextAreaLines: this.startingRows,
    }
  },
  computed: {
    textBoxSize() {
      return 22 * Math.min(this.numTextAreaLines, this.maxRows)
    },
    textBoxOverflow() {
      return this.numTextAreaLines > this.maxRows ? 'auto' : 'hidden'
    },
  },
  watch: {
    value() {
      this.resizeTextArea()
    },
  },
  mounted() {
    this.resizeTextArea()
  },
  methods: {
    resizeTextArea() {
      this.$nextTick(() => {
        const inputTextArea = this.$refs.inputTextArea
        this.numTextAreaLines = this.calculateHeight(inputTextArea)
      })
    },
    /**
     * Taken from https://stackoverflow.com/questions/1760629/how-to-get-number-of-rows-in-textarea-using-javascript/1761203#1761203
     *
     * The key reason we need this is to resize a fully expanded textarea to something
     * smaller as a user deletes newlines or text. Hence we need to actually manipulate
     * the dom and lower the height of the textarea until we find overflow is occurring
     * again to re-find the correct min height.
     */
    calculateContentHeight(ta, scanAmount) {
      const origHeight = ta.style.height
      let height = ta.offsetHeight
      const scrollHeight = ta.scrollHeight
      const overflow = ta.style.overflow
      /// only bother if the ta is bigger than content
      if (height >= scrollHeight) {
        /// check that our browser supports changing dimension
        /// calculations mid-way through a function call...
        ta.style.height = height + scanAmount + 'px'
        /// because the scrollbar can cause calculation problems
        ta.style.overflow = 'hidden'
        /// by checking that scrollHeight has updated
        if (scrollHeight < ta.scrollHeight) {
          /// now try and scan the ta's height downwards
          /// until scrollHeight becomes larger than height
          while (ta.offsetHeight >= ta.scrollHeight) {
            ta.style.height = (height -= scanAmount) + 'px'
          }
          /// be more specific to get the exact height
          while (ta.offsetHeight < ta.scrollHeight) {
            ta.style.height = height++ + 'px'
          }
          height--
          /// reset the ta back to it's original height
          ta.style.height = origHeight
          /// put the overflow back
          ta.style.overflow = overflow
          return height
        } else {
          // We weren't able to resize this element inside this function call to figure
          // out it's true height so we give up and just return the scrollHeight.
          return scrollHeight
        }
      } else {
        return scrollHeight
      }
    },
    calculateHeight(ta) {
      const style = window.getComputedStyle
        ? window.getComputedStyle(ta)
        : ta.currentStyle

      // This will get the line-height only if it is set in the css,
      // otherwise it's "normal"
      const taLineHeight = parseInt(style.lineHeight, 10)
      // Get the scroll height of the textarea
      const taHeight = this.calculateContentHeight(ta, taLineHeight)
      // calculate the number of lines
      return Math.ceil(taHeight / taLineHeight)
    },
  },
}
</script>
