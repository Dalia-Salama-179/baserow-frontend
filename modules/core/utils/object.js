/**
 * Clones the provided Javascript object and returns that one.
 *
 * @param o
 * @return {object}
 */
export function clone(o) {
  return JSON.parse(JSON.stringify(o))
}

/**
 * Creates an object where the key indicates the line number and the value is
 * the string that must be shown on that line number. The line number matches
 * the line number if the value would be stringified with an indent of 4
 * characters `JSON.stringify(value, null, 4)`. The correct value is matched
 * if a value (recursive) object key matches a key of the mapping. However values will
 * not be matched inside the children of a matching key.
 *
 * Example:
 * mappingToStringifiedJSONLines(
 *  { key_2: 'Value' },
 *  {
 *    key_1: 'A random value',
 *    key_2: 'Another value'
 *  }
 * ) === {
 *   3: 'Value'
 * }
 */
export function mappingToStringifiedJSONLines(
  mapping,
  value,
  index = 1,
  lines = {},
  first = true
) {
  if (Array.isArray(value)) {
    index += 1
    value.forEach((v, i) => {
      index = mappingToStringifiedJSONLines(mapping, v, index, lines, false)
    })
    index += 1
    return first ? lines : index
  } else if (value instanceof Object) {
    index += 1
    Object.keys(value).forEach((k) => {
      let childMapping = mapping
      if (Object.prototype.hasOwnProperty.call(mapping, k)) {
        lines[index] = mapping[k]
        // Only recursively search for more field to line mappings where the current key
        // is not itself the key for a field.
        // For example if this key is a field, then there cannot be any other fields
        // to map within this fields value.
        childMapping = {}
      }
      index = mappingToStringifiedJSONLines(
        childMapping,
        value[k],
        index,
        lines,
        false
      )
    })
    index += 1
    return first ? lines : index
  } else {
    index += 1
    return first ? lines : index
  }
}
