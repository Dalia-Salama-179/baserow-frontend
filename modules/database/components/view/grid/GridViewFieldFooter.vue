<template>
  <div
    ref="fieldContextAnchor"
    class="grid-view-aggregation"
    :class="{ 'read-only': readOnly }"
    @click.prevent="
      !readOnly &&
        $refs[`fieldContext`].toggle(
          $refs.fieldContextAnchor,
          'top',
          'right',
          10
        )
    "
  >
    <component
      :is="viewAggregationType.getComponent()"
      v-if="viewAggregationType"
      :loading="loading"
      :aggregation-type="viewAggregationType"
      :value="value"
    />
    <Context :ref="`fieldContext`">
      <ul class="select__items">
        <li
          class="select__item select__item--no-options"
          :class="{ active: !viewAggregationType }"
        >
          <a class="select__item-link" @click="selectAggregation('none')"
            ><div class="select__item-name">{{ $t('common.none') }}</div></a
          >
        </li>
        <li
          v-for="viewAggregation in viewAggregationTypes"
          :key="viewAggregation.getType()"
          class="select__item select__item--no-options"
          :class="{
            active: viewAggregation === viewAggregationType,
          }"
        >
          <a
            class="select__item-link"
            @click="selectAggregation(viewAggregation.getType())"
            ><div class="select__item-name">
              {{ viewAggregation.getName() }}
            </div></a
          >
        </li>
      </ul>
    </Context>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
  name: 'GridViewFieldFooter',
  props: {
    field: {
      type: Object,
      required: true,
    },
    view: {
      type: Object,
      required: true,
    },
    readOnly: {
      type: Boolean,
      required: false,
      default: false,
    },
    storePrefix: {
      type: String,
      required: true,
    },
  },
  computed: {
    aggregationType() {
      return this.fieldOptions[this.field.id]?.aggregation_type
    },
    value() {
      if (this.fieldAggregationData[this.field.id] !== undefined) {
        const { value } = this.fieldAggregationData[this.field.id]

        return this.viewAggregationType.getValue(value, {
          rowCount: this.rowCount,
          field: this.field,
          fieldType: this.fieldType,
        })
      } else {
        return undefined
      }
    },
    loading() {
      return this.fieldAggregationData[this.field.id]?.loading
    },
    viewAggregationType() {
      if (!this.aggregationType) {
        return null
      }
      return this.$registry.get('viewAggregation', this.aggregationType)
    },
    viewAggregationTypes() {
      return this.$registry
        .getOrderedList('viewAggregation')
        .filter((agg) => agg.fieldIsCompatible(this.field))
    },
    fieldType() {
      return this.$registry.get('field', this.field.type)
    },
  },
  watch: {
    aggregationType(value) {
      if (!value) {
        return
      }
      this.$store.dispatch(
        this.storePrefix + 'view/grid/fetchFieldAggregationData',
        {
          view: this.view,
          fieldId: this.field.id,
          options: {
            aggregation_raw_type: this.viewAggregationType.getRawType(),
            aggregation_type: this.viewAggregationType.getType(),
          },
        }
      )
    },
  },
  beforeCreate() {
    this.$options.computed = {
      ...(this.$options.computed || {}),
      ...mapGetters({
        fieldAggregationData:
          this.$options.propsData.storePrefix +
          'view/grid/getAllFieldAggregationData',
        fieldOptions:
          this.$options.propsData.storePrefix + 'view/grid/getAllFieldOptions',
        rowCount: this.$options.propsData.storePrefix + 'view/grid/getCount',
      }),
    }
  },
  methods: {
    async selectAggregation(newType) {
      this.$refs.fieldContext.hide()

      const values = {
        aggregation_type: '',
        aggregation_raw_type: '',
      }

      if (newType !== 'none') {
        const selectedAggregation = this.$registry.get(
          'viewAggregation',
          newType
        )
        values.aggregation_type = newType
        values.aggregation_raw_type = selectedAggregation.getRawType()
      }

      await this.$store.dispatch(
        this.storePrefix + 'view/grid/updateFieldOptionsOfField',
        {
          field: this.field,
          values,
        }
      )
    },
  },
}
</script>
