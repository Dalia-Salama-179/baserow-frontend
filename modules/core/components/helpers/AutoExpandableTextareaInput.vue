<template>
  <div
    class="auto-expandable-textarea__container"
    :class="{
      'auto-expandable-textarea--loading-overlay': loading,
    }"
  >
    <div v-if="loading" class="auto-expandable-textarea--loading"></div>
    <AutoExpandableTextarea
      v-model="innerValue"
      :placeholder="placeholder"
      :max-rows="maxRows"
      :starting-rows="startingRows"
      @keydown.enter.exact.prevent="$emit('entered')"
    >
    </AutoExpandableTextarea>
  </div>
</template>
<script>
import AutoExpandableTextarea from '@baserow/modules/core/components/helpers/AutoExpandableTextarea'

/**
 * A textarea styled as an input box which will automatically expand given the users
 * input to the maximum number of rows specified by the maxRows prop. Prevents enter
 * by default and instead emits an 'entered' event for the parent to handle. Also shows
 * a loading overlay when the loading prop is set to true.
 */
export default {
  name: 'AutoExpandableTextareaInput',
  components: {
    AutoExpandableTextarea,
  },
  props: {
    value: {
      required: true,
      type: String,
    },
    loading: {
      required: false,
      default: false,
      type: Boolean,
    },
    placeholder: {
      required: false,
      type: String,
      default: '',
    },
    maxRows: {
      required: false,
      type: Number,
      default: 4,
    },
    startingRows: {
      required: false,
      type: Number,
      default: 1,
    },
  },
  computed: {
    innerValue: {
      get() {
        return this.value
      },
      set(value) {
        this.$emit('input', value)
      },
    },
  },
}
</script>
