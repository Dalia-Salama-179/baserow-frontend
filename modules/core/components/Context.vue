<template>
  <div
    class="context"
    :class="{ 'visibility-hidden': !open || !updatedOnce }"
    @click="onClick($event)"
  >
    <slot v-if="openedOnce"></slot>
  </div>
</template>

<script>
import { isElement, isDomElement } from '@baserow/modules/core/utils/dom'
import MoveToBody from '@baserow/modules/core/mixins/moveToBody'

export default {
  name: 'Context',
  mixins: [MoveToBody],
  props: {
    hideOnClickOutside: {
      type: Boolean,
      default: true,
      required: false,
    },
  },
  data() {
    return {
      open: false,
      opener: null,
      updatedOnce: false,
      // If opened once, should stay in DOM to keep nested content
      openedOnce: false,
      insideEvent: new Set(),
    }
  },
  methods: {
    /**
     * Toggles the open state of the context menu.
     *
     * @param target      The original element element that changed the state of the
     *                    context, this will be used to calculate the correct position.
     * @param vertical    Bottom positions the context under the target.
     *                    Top positions the context above the target.
     *                    Over-bottom positions the context over and under the target.
     *                    Over-top positions the context over and above the target.
     * @param horizontal  `left` aligns the context with the left side of the target.
     *                    `right` aligns the context with the right side of the target.
     * @param verticalOffset
     *                    The offset indicates how many pixels the context is moved
     *                    top from the original calculated position.
     * @param horizontalOffset
     *                    The offset indicates how many pixels the context is moved
     *                    left from the original calculated position.
     * @param value       True if context must be shown, false if not and undefine
     *                    will invert the current state.
     */
    toggle(
      target,
      vertical = 'bottom',
      horizontal = 'left',
      verticalOffset = 10,
      horizontalOffset = 0,
      value
    ) {
      if (value === undefined) {
        value = !this.open
      }

      if (value) {
        this.show(
          target,
          vertical,
          horizontal,
          verticalOffset,
          horizontalOffset
        )
      } else {
        this.hide()
      }
    },
    /**
     * Add the event to the `insideEvent` map. This allow to be sure a click event has
     * been triggered from an element inside this context, even if the element has
     * been removed after in the meantime.
     */
    onClick(event) {
      this.insideEvent.add(event)
    },
    /**
     * Calculate the position, show the context menu and register a click event on the
     * body to check if the user has clicked outside the context.
     */
    async show(
      target,
      vertical,
      horizontal,
      verticalOffset = 10,
      horizontalOffset = 0
    ) {
      const isElementOrigin = isDomElement(target)
      const updatePosition = () => {
        const css = isElementOrigin
          ? this.calculatePositionElement(
              target,
              vertical,
              horizontal,
              verticalOffset,
              horizontalOffset
            )
          : this.calculatePositionFixed(
              target,
              vertical,
              horizontal,
              verticalOffset,
              horizontalOffset
            )

        // Set the calculated positions of the context.
        for (const key in css) {
          const value = css[key] !== null ? Math.ceil(css[key]) + 'px' : 'auto'
          this.$el.style[key] = value
        }
        this.updatedOnce = true
      }

      // If we store the element who opened the context menu we can exclude the element
      // when clicked outside of this element.
      this.opener = isElementOrigin ? target : null

      this.open = true
      this.openedOnce = true

      // Delay the position update to the next tick to let the Context content
      // be available in DOM for accurate positioning.
      await this.$nextTick()
      updatePosition()

      this.$el.clickOutsideEvent = (event) => {
        // If the event is from current context or any element inside current context
        // the current event should be in the insideEvent map, even if the element
        // has been removed from the DOM in the meantime
        const insideContext = this.insideEvent.has(event)
        if (insideContext) {
          this.insideEvent.delete(event)
        }

        // Check if the context menu is still open
        if (this.open) {
          if (
            // If the prop allows it to be closed by clicking outside.
            this.hideOnClickOutside &&
            // If the click was outside the context element because we want to ignore
            // clicks inside it or any child of this element
            !insideContext &&
            // If the click was not on the opener because he can trigger the toggle
            // method.
            !isElement(this.opener, event.target) &&
            // If the click was not inside one of the context children of this context
            // menu.
            !this.moveToBody.children.some((child) => {
              return isElement(child.$el, event.target)
            })
          ) {
            this.hide()
          }
        }
      }
      document.body.addEventListener('click', this.$el.clickOutsideEvent)

      this.$el.updatePositionEvent = (event) => {
        updatePosition()
      }
      window.addEventListener('scroll', this.$el.updatePositionEvent, true)
      window.addEventListener('resize', this.$el.updatePositionEvent)

      this.$emit('shown')
    },
    /**
     * Forces the child elements to render by setting `openedOnce` to `true`. This
     * could be useful when children of the context must be accessed before the context
     * has been opened.
     */
    forceRender() {
      this.openedOnce = true
    },
    /**
     * Hide the context menu and make sure the body event is removed.
     */
    hide(emit = true) {
      this.opener = null
      this.open = false

      if (emit) {
        this.$emit('hidden')
      }

      this.insideEvent = new Set()

      document.body.removeEventListener('click', this.$el.clickOutsideEvent)
      window.removeEventListener('scroll', this.$el.updatePositionEvent, true)
      window.removeEventListener('resize', this.$el.updatePositionEvent)
    },
    /**
     * Calculates the absolute position of the context based on the original clicked
     * element. If the target element is not visible, it might mean that we can't
     * figure out the correct position, so in that case we force the element to be
     * visible.
     */
    calculatePositionElement(
      target,
      vertical,
      horizontal,
      verticalOffset,
      horizontalOffset
    ) {
      const visible =
        window.getComputedStyle(target).getPropertyValue('display') !== 'none'

      // If the target is not visible then we can't calculate the position, so we
      // temporarily need to show the element forcefully.
      if (!visible) {
        target.classList.add('forced-block')
      }

      const targetRect = target.getBoundingClientRect()
      const contextRect = this.$el.getBoundingClientRect()
      const positions = { top: null, right: null, bottom: null, left: null }

      if (!visible) {
        target.classList.remove('forced-block')
      }

      // Take into account that the document might be scrollable.
      verticalOffset += document.documentElement.scrollTop
      horizontalOffset += document.documentElement.scrollLeft

      // Calculate if top, bottom, left and right positions are possible.
      const canTop = targetRect.top - contextRect.height - verticalOffset > 0
      const canBottom =
        window.innerHeight -
          targetRect.bottom -
          contextRect.height -
          verticalOffset >
        0
      const canOverTop =
        targetRect.bottom - contextRect.height - verticalOffset > 0
      const canOverBottom =
        window.innerHeight -
          targetRect.bottom -
          contextRect.height -
          verticalOffset >
        0
      const canRight =
        targetRect.right - contextRect.width - horizontalOffset > 0
      const canLeft =
        window.innerWidth -
          targetRect.left -
          contextRect.width -
          horizontalOffset >
        0

      // If bottom, top, left or right doesn't fit, but their opposite does we switch to
      // that.
      if (vertical === 'bottom' && !canBottom && canTop) {
        vertical = 'top'
      }

      if (vertical === 'top' && !canTop) {
        vertical = 'bottom'
      }

      if (vertical === 'over-bottom' && !canOverBottom && canOverTop) {
        vertical = 'over-top'
      }

      if (vertical === 'over-top' && !canOverTop) {
        vertical = 'over-bottom'
      }

      if (horizontal === 'left' && !canLeft && canRight) {
        horizontal = 'right'
      }

      if (horizontal === 'right' && !canRight) {
        horizontal = 'left'
      }

      // Calculate the correct positions for horizontal and vertical values.
      if (horizontal === 'left') {
        positions.left = targetRect.left + horizontalOffset
      }

      if (horizontal === 'right') {
        positions.right =
          window.innerWidth - targetRect.right - horizontalOffset
      }

      if (vertical === 'bottom') {
        positions.top = targetRect.bottom + verticalOffset
      }

      if (vertical === 'top') {
        positions.bottom = window.innerHeight - targetRect.top + verticalOffset
      }

      if (vertical === 'over-bottom') {
        positions.top = targetRect.top + verticalOffset
      }

      if (vertical === 'over-top') {
        positions.bottom =
          window.innerHeight - targetRect.bottom + verticalOffset
      }

      return positions
    },
    /**
     * Calculates the desired position based on the provided coordinates. For now this
     * is only used by the row context menu, but because of the reserved space of the
     * grid on the right and bottom there is always room for the context. Therefore we
     * do not need to check if the context fits.
     */
    calculatePositionFixed(coordinates) {
      return {
        left: coordinates.left,
        top: coordinates.top,
        right: null,
        bottom: null,
      }
    },
  },
}
</script>
