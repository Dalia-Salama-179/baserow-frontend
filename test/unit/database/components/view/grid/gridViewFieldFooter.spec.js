import { TestApp } from '@baserow/test/helpers/testApp'
import GridViewFieldFooter from '@baserow/modules/database/components/view/grid/GridViewFieldFooter'
import Context from '@baserow/modules/core/components/Context'
import { clone } from '@baserow/modules/core/utils/object'
import flushPromises from 'flush-promises'

describe('Field footer component', () => {
  let testApp = null
  let mockServer = null
  let store = null

  beforeAll(() => {
    testApp = new TestApp()
    store = testApp.store
    mockServer = testApp.mockServer
  })

  afterEach((done) => {
    testApp.afterEach().then(done)
  })

  const mountComponent = (props, slots = {}) => {
    return testApp.mount(GridViewFieldFooter, { propsData: props, slots })
  }

  const selectValue = async (wrapper, child) => {
    await wrapper.find(`.grid-view-aggregation`).trigger('click')
    const context = wrapper.findComponent(Context)

    await context
      .find(`.select__items > .select__item:nth-child(${child})`)
      .find('.select__item-link')
      .trigger('click')
  }

  test('Default component', async () => {
    await store.dispatch('page/view/grid/forceUpdateAllFieldOptions', {
      2: {
        aggregation_type: 'not_empty_percentage',
        aggregation_raw_type: 'not_empty_count',
      },
    })

    store.commit('page/view/grid/SET_COUNT', 1024)

    const view = {
      id: 1,
    }

    // field with no aggregation
    const wrapper1 = await mountComponent({
      view,
      field: { id: 1, type: 'text' },
      storePrefix: 'page/',
    })
    expect(wrapper1.element).toMatchSnapshot()

    // Field with aggregation
    const wrapper2 = await mountComponent({
      view,
      field: { id: 2, type: 'text' },
      storePrefix: 'page/',
    })

    expect(wrapper2.element).toMatchSnapshot()

    mockServer.getAllFieldAggregationData(view.id, {
      field_2: 256,
    })

    // let's fetch the data for this field
    await store.dispatch('page/view/grid/fetchAllFieldAggregationData', {
      view,
    })

    expect(wrapper2.element).toMatchSnapshot()
  })

  test('Change type', async () => {
    await store.dispatch('page/view/grid/forceUpdateAllFieldOptions', {
      3: {
        aggregation_type: 'not_empty_count',
        aggregation_raw_type: 'not_empty_count',
      },
    })

    store.commit('page/view/grid/SET_LAST_GRID_ID', 2)

    const view = {
      id: 2,
    }

    mockServer.getAllFieldAggregationData(view.id, {
      field_3: 256,
    })
    mockServer.updateFieldOptions(view.id, {
      3: {
        aggregation_type: '',
        aggregation_raw_type: '',
      },
    })

    // Field with aggregation
    const wrapper = await mountComponent({
      view,
      field: { id: 3, type: 'text' },
      storePrefix: 'page/',
    })

    // let's fetch the data for this field
    await store.dispatch('page/view/grid/fetchAllFieldAggregationData', {
      view,
    })

    // Open menu manually first to have the opportunity to make snapshots
    await wrapper.find(`.grid-view-aggregation`).trigger('click')
    const context = wrapper.findComponent(Context)

    expect(context.element).toMatchSnapshot()

    // Click on aggregation type empty_count
    await context
      .find('.select__items > .select__item:nth-child(1)')
      .find('.select__item-link')
      .trigger('click')

    await flushPromises()

    expect(wrapper.element).toMatchSnapshot()

    mockServer.getAllFieldAggregationData(view.id, {
      field_3: 10,
    })
    mockServer.updateFieldOptions(view.id, {
      3: {
        aggregation_type: 'empty_count',
        aggregation_raw_type: 'empty_count',
      },
    })

    await store.dispatch('page/view/grid/forceUpdateAllFieldOptions', {
      3: {
        aggregation_type: 'empty_count',
        aggregation_raw_type: 'empty_count',
      },
    })

    // Select empty count aggregation now
    await selectValue(wrapper, 2)

    await flushPromises()

    expect(
      clone(store.getters['page/view/grid/getAllFieldAggregationData'])
    ).toEqual({ 3: { loading: false, value: 10 } })

    expect(wrapper.element).toMatchSnapshot()
  })
})
