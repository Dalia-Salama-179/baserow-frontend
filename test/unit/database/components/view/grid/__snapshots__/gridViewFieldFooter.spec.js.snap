// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field footer component Change type 1`] = `
<div
  class="context visibility-hidden"
  style="top: 10px; left: 0px;"
>
  <ul
    class="select__items"
  >
    <li
      class="select__item select__item--no-options"
    >
      <a
        class="select__item-link"
      >
        <div
          class="select__item-name"
        >
          common.none
        </div>
      </a>
    </li>
     
    <li
      class="select__item select__item--no-options"
    >
      <a
        class="select__item-link"
      >
        <div
          class="select__item-name"
        >
          
            viewAggregationType.emptyCount
          
        </div>
      </a>
    </li>
    <li
      class="select__item select__item--no-options active"
    >
      <a
        class="select__item-link"
      >
        <div
          class="select__item-name"
        >
          
            viewAggregationType.notEmptyCount
          
        </div>
      </a>
    </li>
    <li
      class="select__item select__item--no-options"
    >
      <a
        class="select__item-link"
      >
        <div
          class="select__item-name"
        >
          
            viewAggregationType.emptyPercentage
          
        </div>
      </a>
    </li>
    <li
      class="select__item select__item--no-options"
    >
      <a
        class="select__item-link"
      >
        <div
          class="select__item-name"
        >
          
            viewAggregationType.notEmptyPercentage
          
        </div>
      </a>
    </li>
    <li
      class="select__item select__item--no-options"
    >
      <a
        class="select__item-link"
      >
        <div
          class="select__item-name"
        >
          
            viewAggregationType.uniqueCount
          
        </div>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Field footer component Change type 2`] = `
<div
  class="grid-view-aggregation"
>
  <!---->
   
</div>
`;

exports[`Field footer component Change type 3`] = `
<div
  class="grid-view-aggregation"
>
  <div
    class="grid-view-aggregation__generic"
  >
    <span
      class="grid-view-aggregation__generic__name"
    >
      viewAggregationType.emptyCount
    </span>
     
    <span
      class="grid-view-aggregation__generic__value"
    >
      10
    </span>
  </div>
   
</div>
`;

exports[`Field footer component Default component 1`] = `
<div
  class="grid-view-aggregation"
>
  <!---->
   
</div>
`;

exports[`Field footer component Default component 2`] = `
<div
  class="grid-view-aggregation"
>
  <div
    class="grid-view-aggregation__generic"
  >
    <span
      class="grid-view-aggregation__generic__name"
    >
      viewAggregationType.notEmptyCount
    </span>
     
    <span
      class="grid-view-aggregation__generic__value"
    >
      
    </span>
  </div>
   
</div>
`;

exports[`Field footer component Default component 3`] = `
<div
  class="grid-view-aggregation"
>
  <div
    class="grid-view-aggregation__generic"
  >
    <span
      class="grid-view-aggregation__generic__name"
    >
      viewAggregationType.notEmptyCount
    </span>
     
    <span
      class="grid-view-aggregation__generic__value"
    >
      75%
    </span>
  </div>
   
</div>
`;
